src/Microsoft.Data.Analysis/StringDataFrameColumn.cs

```diff
@@ -171,29 +171,73 @@ public IEnumerator<string> GetEnumerator()
+            PrimitiveDataFrameColumn<long> columnSortIndices;
+            if (ascending)
+            {
+                columnSortIndices = GetAscendingSortIndices();
+            }
+            else
+            {
+                columnSortIndices = GetDescendingSortIndices();
+            }
+            return Clone(columnSortIndices);
         }
 
         internal override PrimitiveDataFrameColumn<long> GetAscendingSortIndices()
         {
-            GetSortIndices(Comparer<string>.Default, out PrimitiveDataFrameColumn<long> columnSortIndices);
-            return columnSortIndices;
+            PrimitiveDataFrameColumn<long> sortIndices = GetSortIndices(Comparer<string>.Default, out PrimitiveDataFrameColumn<long> columnNullIndices);
+            if(columnNullIndices.Length > 0)
+            {
+                for (int i=0; i< columnNullIndices.Length; i++)
+                {
+                    sortIndices.Append(columnNullIndices[i]);
+                }
+            }
+            return sortIndices;
         }
 
-        private void GetSortIndices(Comparer<string> comparer, out PrimitiveDataFrameColumn<long> columnSortIndices)
+        internal override PrimitiveDataFrameColumn<long> GetDescendingSortIndices()
+        {
+            Comparison<string> descendingComparison = new Comparison<string>((string a, string b) =>
+            {
+                return Comparer<string>.Default.Compare(b, a);
+            });
+            Comparer<string> descending = Comparer<string>.Create(descendingComparison);
+            // The return sortIndices contains only the non null indices. 
+            PrimitiveDataFrameColumn<long> sortIndices = GetSortIndices(descending, out PrimitiveDataFrameColumn<long> nullIndices);
+            for (long i = 0; i < nullIndices.Length; i++)
+            {
+                sortIndices.Append(nullIndices[i]);
+            }
+            return sortIndices;
+        }
+
+        private PrimitiveDataFrameColumn<long> GetSortIndices(Comparer<string> comparer, out PrimitiveDataFrameColumn<long> columnNullIndices)
         {
             List<int[]> bufferSortIndices = new List<int[]>(_stringBuffers.Count);
+            columnNullIndices = new PrimitiveDataFrameColumn<long>("columnNullIndices");
+            
             foreach (List<string> buffer in _stringBuffers)
             {
                 var sortIndices = new int[buffer.Count];
+
+
                 for (int i = 0; i < buffer.Count; i++)
                 {
                     sortIndices[i] = i;
                 }
                 // TODO: Refactor the sort routine to also work with IList?
                 string[] array = buffer.ToArray();
                 IntrospectiveSort(array, array.Length, sortIndices, comparer);
+                // do a for loop here and get the nonnulls together
+                for(int i=0; i<sortIndices.Length; i++)
+                {
+                    if (buffer[i] is null)
```

Comment:
Shouldn't this be `buffer[sortIndices[i]]` (similar change in the `Append` below)?

Reply:
I don't think that will work either. The issue is there are multiple `_stringBuffers`, and `i` is relative to the current buffer. So we need to account for how many buffers was before this buffer. Just like we do in PrimitiveDataFrameColumn.GetSortIndices.

---
In reply to: [501318342](https://github.com/dotnet/corefxlab/pull/2952#discussion_r501318342) [](ancestors = 501318342)

Reply:
Fixed this locally.